bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManagerTest.java

```diff
@@ -0,0 +1,244 @@
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * Unit test {@link ZkLedgerUnderreplicationManager}.
+ */
+public class ZkLedgerUnderreplicationManagerTest extends BookKeeperClusterTestSuite {
+
+    @BeforeClass
+    public static void setUpCluster() throws Exception {
+        BookKeeperClusterTestSuite.setUpCluster(0);
+    }
+
+    @AfterClass
+    public static void tearDownCluster() throws Exception {
+        BookKeeperClusterTestSuite.tearDownCluster();
+    }
+
+    private static String getZooKeeperConnectString() throws Exception {
+        String[] serviceHosts = metadataStore.getServiceUri().getServiceHosts();
+        return StringUtils.join(serviceHosts, ',');
+    }
+
+    private static ZooKeeper createZooKeeper() throws Exception {
+        return ZooKeeperClient.newBuilder()
+            .connectString(getZooKeeperConnectString())
+            .connectRetryPolicy(
+                new BoundExponentialBackoffRetryPolicy(1, 10, 100))
+            .operationRetryPolicy(
+                new BoundExponentialBackoffRetryPolicy(1, 10, 100))
+            .sessionTimeoutMs(60000)
+            .build();
+    }
+
+    private ZooKeeper zk;
+    private ZkLedgerUnderreplicationManager urMgr;
+
+    @Before
+    public void setUp() throws Exception {
+        this.zk = createZooKeeper();
+        ServerConfiguration conf = new ServerConfiguration(baseServerConf);
+        conf.setStoreSystemTimeAsLedgerUnderreplicatedMarkTime(true);
+        this.urMgr = new ZkLedgerUnderreplicationManager(conf, zk);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        if (null != urMgr) {
+            this.urMgr.close();
+        }
+        if (null != zk) {
+            this.zk.close();
+        }
+    }
+
+    /**
+     * Test basic operation on {@link ZkLedgerUnderreplicationManager#markLedgerUnderreplicatedAsync(long, Collection)}.
+     */
+    @Test
+    public void testMarkLedgerUnderreplicatedBasic() throws Exception {
+        long ledgerId = 0xabcdef;
+        Collection<String> missingBookies = Lists.newArrayList("bookie-1");
+
+        long prevCtime = -1L;
+
+        // mark when it hasn't been marked before
+        FutureUtils.result(urMgr.markLedgerUnderreplicatedAsync(ledgerId, missingBookies));
+        UnderreplicatedLedgerFormat urLedgerFormat = urMgr.getLedgerUnreplicationInfo(ledgerId);
+        assertEquals(missingBookies, urLedgerFormat.getReplicaList());
+        assertTrue(urLedgerFormat.getCtime() > prevCtime);
+        prevCtime = urLedgerFormat.getCtime();
+
+        // mark when it has been marked. but the missing bookies already duplicated there
+        FutureUtils.result(urMgr.markLedgerUnderreplicatedAsync(ledgerId, missingBookies));
+        urLedgerFormat = urMgr.getLedgerUnreplicationInfo(ledgerId);
+        assertEquals(missingBookies, urLedgerFormat.getReplicaList());
+        assertTrue(urLedgerFormat.getCtime() >= prevCtime);
```

Comment:
why are you checking ">=", it should be '==', ctime of marked underreplicated ledger should not change by adding/removing replicaslist to urledger. This would be nice test case to validate that statement.